File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int k = 0;
        int pro = 1;
        int zeroCount = 0;
        for(int i : nums){
            if(i == 0) zeroCount++;
            else pro*=i;
        }

        int ret[] = new int[nums.length];
        for(int i : nums){
            if(i == 0){
                if(zeroCount > 1) ret[k++] = 0;
                else ret[k++] = pro;
            }else{
                if(zeroCount > 0) ret[k++] = 0;
                else ret[k++] = (pro/i);
            }
        }
        return ret;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        isZero = False
        overallProd = 1
        length = len(nums)

        for i in nums:
            if i!=0:
                overallProd *= i
            else:
                if isZero:
                    return [0]*length
                isZero = True
                    
        
        lt = []
        for i in nums:
            if isZero:
                if i == 0:
                    lt.append(overallProd)
                else:
                    lt.append(0)
            else:
                lt.append(overallProd//i)
        
        return lt